Implement IOLoop.run_in_executor

mivade opened the pull request:
Addresses #1982: implements an API mimicking `run_in_executor` from `asyncio`. This uses the same strategy as `asyncio`, namely using `chain_future` to map `concurrent.futures.Future` onto an awaitable `Future`.

Some questions:

- Do we need to do something else to properly handle exceptions here?
- What additional tests should be added?
- Would it make sense to always return the chained future instead of checking the Python version?

bdarnell commented on tornado/ioloop.py:
```diff
@@ -625,6 +632,35 @@ def add_future(self, future, callback):
+            executor = self._executor
+
+        future = executor.submit(func, *args)
+        self.add_future(future, lambda future: future.result())
```
We shouldn't call `add_future` here. That will log the future's result even if the caller handles it.

bdarnell commented on tornado/ioloop.py:
```diff
@@ -625,6 +632,35 @@ def add_future(self, future, callback):
+
+        if executor is None:
+            if not hasattr(self, '_executor'):
+                self._executor = ThreadPoolExecutor(max_workers=(cpu_count() * 5))
```
If `self._executor` is set, it should be closed in `close()`.

mivade replied:
Thanks, good point.

bdarnell commented on tornado/ioloop.py:
```diff
@@ -625,6 +632,35 @@ def add_future(self, future, callback):
+
+        future = executor.submit(func, *args)
+        self.add_future(future, lambda future: future.result())
+        if sys.version_info >= (3, 5):
```
Use the same pattern unconditionally. It's possible to use async/await on older python versions via cython. 

But `chain_future` isn't safe for this cross-thread use case. It needs to be something like `self.add_future(future, lambda f: chain_future(future, future_))`. There should be a test that a callback added to a future returned by `run_in_executor` resolves on the `IOLoop` thread.

bdarnell commented on tornado/test/ioloop_test.py:
```diff
@@ -584,6 +585,39 @@ def handle_exception(typ, value, traceback):
+
+        @gen.coroutine
+        def main():
+            arg1, arg2 = random.random(), random.random()
```
Why introduce randomness? Also, comparing floats for equality is not a great idea. I think a better test would be to submit two functions to the executor and make sure that they run in parallel. 

    event1 = threading.Event()
    event2 = threading.Event()
    def callback(self_event, other_event):
        self_event.set()
        other_event.wait()
        return self_event
    res = yield [
        IOLoop.current().run_in_executor(None, callback, event1, event2),        
        IOLoop.current().run_in_executor(None, callback, event2, event1),
    ]
    self.assertEqual([event1, event2], res)

bdarnell commented on tornado/test/ioloop_test.py:
```diff
@@ -584,6 +585,39 @@ def handle_exception(typ, value, traceback):
+
+        executor = MyExecutor(max_workers=1)
+        IOLoop.current().set_default_executor(executor)
+        self.assertIsInstance(IOLoop.current()._executor, MyExecutor)
```
This doesn't test that `set_default_executor` changes the behavior of `run_in_executor`. Try something like the above test, but instead of `other_event.wait()` do `time.sleep(0.01); self.assertFalse(other_event.is_set())`

bdarnell commented on tornado/ioloop.py:
```diff
@@ -625,6 +632,35 @@ def add_future(self, future, callback):
+        Use `functools.partial` to pass keyword arguments to `func`.
+
+        """
+        if not _concurrent_futures_found:
```
Elsewhere for this pattern we just use `if ThreadPoolExecutor is None` instead of a separate boolean variable.

bdarnell commented on tornado/ioloop.py:
```diff
@@ -43,8 +43,9 @@
 import time
 import traceback
 import math
+from multiprocessing import cpu_count
```
`tornado.process` has a more robust wrapper around `cpu_count`, although I'm not sure if it's needed any more (it mainly dates back to when we supported python 2.5).

mivade replied:
Looking at the Python documentation, it says `cpu_count` may raise a `NotImplementedError` without explaining when or why. Since the function in `tornado.process` defaults to 1 if all else fails, I agree it makes more sense to use that.

mivade commented:
Thanks for the feedback, Ben. I've addressed some of your comments in the original pull request with new commits. The others I will address when I am back from a short vacation.

Another thing I thought about that I'd like your opinion on: `tornado.concurrent.run_on_executor` is one keystroke away from `run_in_executor`. For consistency, would it make sense to provide an alias?

bdarnell commented:
The names are similar, but they're very different things. Given the `asyncio` precedent (post-cedent? It came later but has the weight of standardization), I'd rather use `run_in_executor` for this new method and give `run_on_executor` a new name (or maybe deprecate it and replace it with a new decorator. Now that it can use `IOLoop.current().run_in_executor` it doesn't need to rely on the weird `self.executor` pattern)

kordek commented:
Hi @bdarnell @mivade ! What's the status of this? I'm really interested in this functionality, could help if sth is still needed.

bdarnell commented:
@kordek Looks like two of my comments from Jun 3 are still unaddressed (the use of add_future and testing set_default_executor).

mivade commented:
Sorry for the long silence on this. I'll address the remaining two comments within the next few days.

kordek commented:
Good to know, thanks guys!

mivade commented:
@bdarnell, let me know if you have further comments or suggestions.

PatTheMav commented:
Really looking forward to when this lands - building a web app that should work with Tornado's built-in IOLoop as well as asyncio depending on user choice and want to keep loop-specific code as close to nonexistent as possible (except for installation of the loop itself). 👍

bdarnell commented:
Thanks @mivade !
